hbase-client/src/main/java/org/apache/hadoop/hbase/client/trace/TableOperationSpanBuilder.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.client.trace;
+
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.DB_NAME;
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.DB_OPERATION;
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.NAMESPACE_KEY;
+import static org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.TABLE_KEY;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanKind;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Append;
+import org.apache.hadoop.hbase.client.CheckAndMutate;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Increment;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.RegionCoprocessorServiceExec;
+import org.apache.hadoop.hbase.client.Row;
+import org.apache.hadoop.hbase.client.RowMutations;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.trace.HBaseSemanticAttributes.Operation;
+import org.apache.hadoop.hbase.trace.TraceUtil;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Construct {@link io.opentelemetry.api.trace.Span} instances originating from
+ * "table operations" -- the verbs in our public API that interact with data in tables.
+ */
+@InterfaceAudience.Private
+public class TableOperationSpanBuilder implements Supplier<Span> {
+
+  // n.b. The results of this class are tested implicitly by way of the likes of
+  // `TestAsyncTableTracing` and friends.
+
+  private static final String unknown = "UNKNOWN";
+
+  private TableName tableName;
+  private final Map<AttributeKey<?>, Object> attributes = new HashMap<>();
+
+  @Override public Span get() {
+    return build();
+  }
+
+  public TableOperationSpanBuilder setOperation(final Scan scan) {
+    return setOperation(valueFrom(scan));
+  }
+
+  public TableOperationSpanBuilder setOperation(final Row row) {
+    return setOperation(valueFrom(row));
+  }
+
+  public TableOperationSpanBuilder setOperation(final Operation operation) {
+    attributes.put(DB_OPERATION, operation.name());
+    return this;
+  }
+
+  public TableOperationSpanBuilder setTableName(final TableName tableName) {
+    this.tableName = tableName;
+    attributes.put(NAMESPACE_KEY, tableName.getNamespaceAsString());
+    attributes.put(DB_NAME, tableName.getNamespaceAsString());
+    attributes.put(TABLE_KEY, tableName.getNameAsString());
+    return this;
+  }
+
+  @SuppressWarnings("unchecked")
+  public Span build() {
+    final String name = attributes.getOrDefault(DB_OPERATION, unknown)
```

Comment:
We have different types for Scan, in the past, we only traced scanAll, so I named the span as 'scanAll', and leave the name 'scan' to be used in the future. So what is the plan now?

Reply:
My understanding of the spec says that for anything that is a direct user action should have a span name that matches the DB operation. In this patch, I interpret those operations to map to our table data action verbs -- "get", "put", &c. -- basically matching up to our shell interface. "scan" would be another such user action. It's a good point that "scanAll" exists in the java client API but not in the shell API...

I think that `AsyncTable.scanAll` makes sense to use the DB operation name "SCAN", as I have here. I also noticed that `AsyncTable<C>.scan(Scan, C)` does not have a tracing test. I think we would being back the code you used to have, where each client-side call to `scanner.next` is traced. You mentioned that it could result in many thousands of scans, so you removed it. But I think this is the correct way to handle this part of our API. Anyway, the distributed tracing implementations seem to limit the number of scan children per parent to ~256. I think we should emit scans that actually happen, and leave it up to the tracing service to handle truncation or summarization of scan children.

What do you think?

Reply:
I've raised this question over here: https://cloud-native.slack.com/archives/C01QZFGMLQ7/p1638556336052800

Reply:
Filed [HBASE-26545](https://issues.apache.org/jira/browse/HBASE-26545).

Reply:
However the "long" scan is traced, the both `scan` and `scanAll` methods are "scan" operations from the client's perspective, so I think it's find for both of them to have `db.operation="SCAN"`.

Reply:
How can we trace long scan in the Table code? We just return a Scanner to client or users just pass in a Consumer, I do not think we can have a big span to trace them all? It depends on how user call us.

And what you said about the scanner.next, I was not talking about client side scan, I was talking about the server side RegionScanner, a client scan request can lead to thousands or even more of RegionScanner.next if filter is used, which will impact the scan performance a lot, as it could scan all the whole region but return nothing.

Reply:
And we will always have the rpc method to be traced, so even if we do nothing in the scan method, we could still see a lot of rpc spans when scanning. This is true.

Reply:
> And what you said about the scanner.next, I was not talking about client side scan, I was talking about the server side RegionScanner...

Okay, understood. We can discuss that separately.

> And we will always have the rpc method to be traced, so even if we do nothing in the scan method, we could still see a lot of rpc spans when scanning. This is true.

It sounds like we need to open an operation-level span, just to encapsulate all the RPC spans.